compiler/rustc_resolve/src/late/diagnostics.rs

```diff
@@ -1810,19 +1810,17 @@ impl<'a: 'ast, 'ast> LateResolutionVisitor<'a, '_, 'ast> {
         false
     }
 
-    fn let_binding_suggestion(&self, err: &mut Diagnostic, ident_span: Span) -> bool {
+    fn let_binding_suggestion(&mut self, err: &mut Diagnostic, ident_span: Span) -> bool {
         // try to give a suggestion for this pattern: `name = 1`, which is common in other languages
         let mut added_suggestion = false;
-        if let Some(Expr { kind: ExprKind::Assign(lhs, _rhs, _), .. }) = self.diagnostic_metadata.in_assignment &&
+        if let Some(Expr { kind: ExprKind::Assign(lhs, .. ), .. }) = self.diagnostic_metadata.in_assignment &&
             let ast::ExprKind::Path(None, _) = lhs.kind {
                 let sm = self.r.session.source_map();
                 let line_span = sm.span_extend_to_line(ident_span);
                 let ident_name = sm.span_to_snippet(ident_span).unwrap();
                 // HACK(chenyukang): make sure ident_name is at the starting of the line to protect against macros
```

Comment:
does something like `ident_span.from_expansion()` also work instead of checking the snippet?

Reply:
Just find we have a API `span.can_be_used_for_suggestions`, maybe suitable for this.
And the snippet also used to avoid:

```rust
while x = blah {
 ....
}
```

Any better solution?

Reply:
Got it, this code is also a `desugaring`, nice!
Updated the code.